### DISCUSSION TITLE
What's the correct way to deal with unmanaged strings?

### DISCUSSION THREAD

**Discussion by ghost** [microsoft/CsWin32#156] [Q&A]
CsWin32 is generating a `PCWSTR` type for me,

I've kinda stumbled my way into the following code to deal with it:

```cs
unsafe static PCWSTR GetPCWSTRFromString(string input)
{
    fixed (char* p = input)
    {
        return new PCWSTR(p);
    }
}
```

I can also see if I keep working with various Win32 APIs I'll be dealing with PSTR, LPSTR, PCSTR, LPCSTR, PWSTR, LPWSTR, PCWSTR and LPWSTR as well as various char pointers. Where do I go to learn how these map to C#? It's not something I've really encountered (my day job is in webdev). A lot of the examples I'm seeing around the place are "just" using IntPtr as they were written before CsWin32 was a thing.

**Comment by AArnott** [comment]
You shouldn't need that helper method at all. 

First of all, if you hit F12 (go to definition) on `PCWSTR` you'll find that it includes implicit operators, including:

```cs
        public static implicit operator PCWSTR(char *value) => new PCWSTR(value);
```

That allows you to reduce your helper slightly:

```cs
unsafe static PCWSTR GetPCWSTRFromString(string input)
{
    fixed (char* p = input)
    {
        return p;
    }
}
```

But the above retains a critical fault, which is that C# and .NET only pin your string within the `fixed` block. Once you exit, the string can move and your `PCWSTR` is now pointing to bad memory (if it works, you're only getting "lucky"). Your `fixed` statement must surround the entire useful lifespan of the `PCWSTR`, making such helper methods impossible.

But we solve this for you so you don't need it in the first place. Take `ExtractIcon` for example. See the code we generate:

```cs
    internal static unsafe DestroyIconSafeHandle ExtractIcon(HINSTANCE hInst, string pszExeFileName, uint nIconIndex)
    {
        fixed (char *pszExeFileNameLocal = pszExeFileName)
        {
            HICON __result = PInvoke.ExtractIcon(hInst, pszExeFileNameLocal, nIconIndex);
            return new DestroyIconSafeHandle(__result, ownsHandle: true);
        }
    }

    [DllImport("Shell32", ExactSpelling = true, EntryPoint = "ExtractIconW")]
    internal static extern HICON ExtractIcon(HINSTANCE hInst, PCWSTR pszExeFileName, uint nIconIndex);
}
```

See how we generate *two* overloads, and one takes a `string` parameter? See how we already write the `fixed` statement for you and do all the work that needs it *within* the `fixed* block?

So all you need to do is call it:

```cs
HINSTANCE hInstance;
string dllPath;
uint index;
using SafeHandle iconHandle = PInvoke.ExtractIcon(hInstance, dllPath, index);
```

**Reply by ghost** [reply]
I guess I was looking at other code that needed a `HICON` and I couldn't see how that related to a `SafeHandle` or a `Destroy
IconSafeHandle` but this answer has pointed me in the right direction.
